### PR TITLE
dbeaver/pro#2377 App becomes unresponsive when switching router on a certain schema

### PR DESCRIPTION
**Pull request by serjiokov**
Fixed circling graph, enhanced layout
image

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/layout/algorithm/direct/OrthoDirectedGraphLayout.java [hunk 1]
```diff
@@ -225,6 +227,13 @@ private TreeMap<String, List<Node>> recomputeGraph(TreeMap<String, List<Node>> g
         int idx = 0;
         while (idx < graph.keySet().size()) {
             createGraphLayers(graph, idx);
+            // catch empty nodes   
+            List<Node> nextLevelNodes = graph.get(String.valueOf(idx+1));
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/layout/algorithm/direct/OrthoDirectedGraphLayout.java]: Please fix the code style.

Also, could you tell me why you're using integers converted to strings as keys? Is there a particular reason for that? It requires more computations than just boxing that integer.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/layout/algorithm/direct/OrthoDirectedGraphLayout.java [hunk 2]
```diff
@@ -287,7 +295,26 @@ private void createGraphLayers(TreeMap<String, List<Node>> nodeByEdges, int idx)
...
                 if (trg != null) {
                     String nodeIndex = getNodeIndex(nodeByEdges, trg);
+                    // skip by distance
+                    boolean skip = false;
+                    for (Edge e : trg.incoming) {
+                        Node incomingSourceNode = e.source;
+                        String childIndexStr = getNodeIndex(nodeByEdges, incomingSourceNode);
+                        if (!childIndexStr.isEmpty() && !childIndexStr.equals("0")) {
+                            Integer childIndex = Integer.valueOf(childIndexStr);
+                            if ((idx - childIndex) > DISTANCE_BTW_LEVELS) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/layout/algorithm/direct/OrthoDirectedGraphLayout.java:303]: I'm not sure how it works, but will it handle when there are two tables that point to each other?

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/layout/algorithm/direct/OrthoDirectedGraphLayout.java (hunk 1 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/layout/algorithm/direct/OrthoDirectedGraphLayout.java]: Generally yes it is not needed, I think all enhancement will be done in another ticket.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/layout/algorithm/direct/OrthoDirectedGraphLayout.java (hunk 2 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/layout/algorithm/direct/OrthoDirectedGraphLayout.java:303]: If two or more tables will point to each other tables will be on the same level.  This part just handle distance btw entities.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/layout/algorithm/direct/OrthoDirectedGraphLayout.java (hunk 1 above)
Comment (serge-rider) [reply plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/layout/algorithm/direct/OrthoDirectedGraphLayout.java]: Let's fix code style though. 

And generally changing String to Integer is easy, why not do it right here?

**Review by serge-rider — CHANGES_REQUESTED** [review]

**Review by ShadelessFox — APPROVED** [review]

**Comment by uslss** [comment]
verified

**Pull request merged**